SavedInstances/MythicPlus.lua

```diff
@@ -84,3 +87,80 @@ function MythicPlusModule:RefreshMythicKeyInfo(event)
   t.MythicKeyBest.level = C_MythicPlus_GetWeeklyChestRewardLevel()
   t.MythicKeyBest.WeeklyReward = C_MythicPlus_IsWeeklyRewardAvailable()
 end
+
+function MythicPlusModule:Keys()
+  local target = addon.db.Tooltip.KeystoneReportTarget
+  addon.debug("Key report target: %s", target)
+
+  if target == 'EXPORT' then
+    MythicPlusModule:ExportKeys()
+  else
+    local dialog = StaticPopup_Show("SAVEDINSTANCES_REPORT_KEYS", target, nil, target)
+  end
+end
+
+function MythicPlusModule:KeyData(action)
+  local cpairs = addon.cpairs
+
+  for toon, t in cpairs(addon.db.Toons, true) do
+    if t.MythicKey and t.MythicKey.link then
+      local toonname
+      if addon.db.Tooltip.ShowServer then
+        toonname = toon
+      else
+        local tname, tserver = toon:match('^(.*) [-] (.*)$')
+        toonname = tname
+      end
+
+      action(toonname, t.MythicKey.link)
+    end
+  end
+end
+
+function MythicPlusModule:ReportKeys(target)
+  MythicPlusModule:KeyData(function(toon, key) SendChatMessage(toon .. ' - '.. key, target) end)
+end
+
+function MythicPlusModule:ExportKeys()
+  local keys = ""
+
+  MythicPlusModule:KeyData(function(toon, key) keys = keys .. toon .. ' - ' .. key .. '\n' end)
+
+  if not self.SavedInstancesKeyExport then
+    local f = CreateFrame("Frame", "SavedInstancesKeyExport", UIParent, "DialogBoxFrame")
+    f:SetSize(700, 450)
+    f:SetPoint("CENTER")
+    f:SetFrameStrata("HIGH")
+    f:SetMovable(true)
+    f:EnableMouse(true)
+    f:SetClampedToScreen(true)
+    f:RegisterForDrag("LeftButton")
+    f:SetScript("OnDragStart", f.StartMoving)
+    f:SetScript("OnDragStop", f.StopMovingOrSizing)
+    self.SavedInstancesKeyExport = f
+
+    local sf = CreateFrame("ScrollFrame", "SavedInstancesKeyExportScroll", f, "UIPanelScrollFrameTemplate")
+    sf:SetPoint("TOP", 0, -16)
+    sf:SetPoint("BOTTOM", f, "BOTTOM", 0, 50)
+    sf:SetPoint("LEFT", 16, 0)
+    sf:SetPoint("RIGHT", -36, 0)
+    self.SavedInstancesKeyExportScroll = sf
+
+    local st = CreateFrame("EditBox", "SavedInstancesKeyExportScrollText", sf)
+    st:SetSize(sf:GetSize())
+    st:SetMultiLine(true)
+    st:SetFontObject("ChatFontNormal")
+    st:SetScript("OnEscapePressed", function()
+      f:Hide()
+    end)
+    sf:SetScrollChild(st)
+    self.SavedInstancesKeyExportScrollText = st
+  end

```

Comment:
These frames are assigned to `MythicPlusModule`, so I'd prefer don't let it exported to `_G`.

Reply:
Here my total lack of experience with Lua is showing, so I'm not completely sure how to resolve these remaining two.

Quick googling for how `_G` and `self` work in Lua, my proposed solution would be to remove `self` from this code snippet and leave the variable declarations in the other change. From my understanding this would satisfy your demands?

Reply:
No, `self` here is `MythicPlusModule`, you can simply set the second arg of `CreateFrame` to nil, to prevent WoW exporting to `_G`.

```suggestion
  if not self.SavedInstancesKeyExport then
    local f = CreateFrame("Frame", nil, UIParent, "DialogBoxFrame")
    f:SetSize(700, 450)
    f:SetPoint("CENTER")
    f:SetFrameStrata("HIGH")
    f:SetMovable(true)
    f:EnableMouse(true)
    f:SetClampedToScreen(true)
    f:RegisterForDrag("LeftButton")
    f:SetScript("OnDragStart", f.StartMoving)
    f:SetScript("OnDragStop", f.StopMovingOrSizing)
    self.SavedInstancesKeyExport = f

    local sf = CreateFrame("ScrollFrame", nil, f, "UIPanelScrollFrameTemplate")
    sf:SetPoint("TOP", 0, -16)
    sf:SetPoint("BOTTOM", f, "BOTTOM", 0, 50)
    sf:SetPoint("LEFT", 16, 0)
    sf:SetPoint("RIGHT", -36, 0)
    self.SavedInstancesKeyExportScroll = sf

    local st = CreateFrame("EditBox", nil, sf)
    st:SetSize(sf:GetSize())
    st:SetMultiLine(true)
    st:SetFontObject("ChatFontNormal")
    st:SetScript("OnEscapePressed", function()
      f:Hide()
    end)
    sf:SetScrollChild(st)
    self.SavedInstancesKeyExportScrollText = st
  end

```

Reply:
> No, `self` here is `MythicPlusModule`, you can simply set the second arg of `CreateFrame` to nil, to prevent WoW exporting to `_G`.

Ok, so I didn't know about the `CreateFrame()` auto-export thing. But just for the sake of my understanding: if I'd set the second argument to `nil`, it wouldn't matter much if I'd used either an attribute on `self` or a local var? For example:

```Lua
local foo

function Module:Export()
  if not foo then
    foo = CreateFrame("Frame", nil)
  end
end
```

and

```Lua
function Module:Export()
  if not self.foo then
    self.foo = CreateFrame("Frame", nil)
  end
end
```

basically achieve the same thing? And your suggestion of using an attribute on `MythicPlusModule` is just a preference?

Reply:
Yes. Also, currently you use `self.SavedInstancesKeyExport = f`, the `self` here is basically `MythicPlusModule`.

Reply:
I understand that `self` refers to `MythicPlusModule` in that place. Wouldn't it be better to use `self` though, in case you ever rename `MythicPlusModule`?